specs/bep-v1.rst

```diff
@@ -62,468 +62,277 @@ another.
-message. In Response messages it is set to the Message ID of the corresponding
-Request message. The uniqueness requirement implies that no more than 4096
-request messages may be outstanding at any given moment. For message types
-that do not have a corresponding response (Cluster Configuration, Index, etc.)
-the Message ID field is irrelevant and SHOULD be set to zero.
-
-The **Type** field indicates the type of data following the message header
-and is one of the integers defined below. A message of an unknown type
-is a protocol error and MUST result in the connection being terminated.
-
-The **Compression** bit "C" indicates the compression used for the message.
-
-For C=0:
-
--  The Length field contains the length, in bytes, of the uncompressed
-   message data.
-
--  The message is not compressed.
-
-For C=1:
-
--  The Length field contains the length, in bytes, of the compressed
-   message data plus a four byte uncompressed length field.
-
--  The compressed message data is preceded by a 32 bit field denoting
-   the length of the uncompressed message.
-
--  The message data is compressed using the LZ4 format and algorithm
-   described in http://www.lz4.org/.
-
-All data within the message (post decompression, if compression is in
-use) MUST be in XDR (RFC 1014) encoding. All fields shorter than 32 bits
-and all variable length data MUST be padded to a multiple of 32 bits.
-The actual data types in use by BEP, in XDR naming convention, are the
-following:
+Every message post authentication is made up of several parts:
+
+- A header length word
+- A **Header**
+- A message length word
+- A **Message**
+
+.. code-block:: none
+
+     0                   1
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |         Header Length         |
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /                               /
+    \            Header             \
+    /                               /
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |         Message Length        |
+    |           (32 bits)           |
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /                               /
+    \            Message            \
+    /                               /
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+The header length word is 16 bits. It indicates the length of the following
+**Header** message. The Header is in protocol buffer format. The Header
+describes the type and compression status of the following message.
+
+The message is preceded by the 32 bit message length word and is one of the
+concrete BEP messages described below, identified by the **type** field of
+the Header.
+
+As always, the length words are in network byte order (big endian).
+
+.. code-block:: proto
+
+    message Header {
+        MessageType        type        = 1;
+        MessageCompression compression = 2;
+    }
 
-(unsigned) int:
-    (unsigned) 32 bit integer
+    enum MessageType {
+        UNKNOWN           = 0;
```

Comment:
@burkemw3 I added this, but I'm not sure it's necessary after all. Quoting the proto3 spec (https://developers.google.com/protocol-buffers/docs/proto3#enum, a bit further down):

> During deserialization, unrecognized enum values will be preserved in the message, though how this is represented when the message is deserialized is language-dependent. In languages that support open enum types with values outside the range of specified symbols, such as C++ and Go, the unknown enum value is simply stored as its underlying integer representation. In languages with closed enum types such as Java, a case in the enum is used to represent an unrecognized value, and the underlying integer can be accessed with special accessors. In either case, if the message is serialized the unrecognized value will still be serialized with the message.

So how undefined values are represented depends on the language, but it doesn't seem that it ever folds to the default. As such is seems cleaner to not have `UNKNOWN` here?